<a href="https://colab.research.google.com/github/Ramyavenkatrajulu/Ramyavenkatrajulu/blob/main/Assigment3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [ ]:
users = pd.read_csv('/content/takehome_users.csv',encoding='ISO-8859-1')
print(users.shape)
users.head()

(12000, 10)


object_id        creation_time               name  \
0          1  2014-04-22 03:53:30     Clausen August   
1          2  2013-11-15 03:45:04      Poole Matthew   
2          3  2013-03-19 23:14:52  Bottrill Mitchell   
3          4  2013-05-21 08:09:28    Clausen Nicklas   
4          5  2013-01-17 10:14:20          Raw Grace   

                        email creation_source  last_session_creation_time  \
0    AugustCClausen@yahoo.com    GUEST_INVITE                1.398139e+09   
1      MatthewPoole@gustr.com      ORG_INVITE                1.396238e+09   
2  MitchellBottrill@gustr.com      ORG_INVITE                1.363735e+09   
3   NicklasSClausen@yahoo.com    GUEST_INVITE                1.369210e+09   
4          GraceRaw@yahoo.com    GUEST_INVITE                1.358850e+09   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         1                           0      11   
1                         0                           0       1   
2                         0                           0      94   
3                         0                           0       1   
4                         0                           0     193   

   invited_by_user_id  
0             10803.0  
1               316.0  
2              1525.0  
3              5151.0  
4              5240.0

In [ ]:
users.describe().T

count          mean           std           min  \
object_id                   12000.0  6.000500e+03  3.464246e+03  1.000000e+00   
last_session_creation_time   8823.0  1.379279e+09  1.953116e+07  1.338452e+09   
opted_in_to_mailing_list    12000.0  2.495000e-01  4.327417e-01  0.000000e+00   
enabled_for_marketing_drip  12000.0  1.493333e-01  3.564316e-01  0.000000e+00   
org_id                      12000.0  1.418846e+02  1.240567e+02  0.000000e+00   
invited_by_user_id           6417.0  5.962957e+03  3.383762e+03  3.000000e+00   

                                     25%           50%           75%  \
object_id                   3.000750e+03  6.000500e+03  9.000250e+03   
last_session_creation_time  1.363195e+09  1.382888e+09  1.398443e+09   
opted_in_to_mailing_list    0.000000e+00  0.000000e+00  0.000000e+00   
enabled_for_marketing_drip  0.000000e+00  0.000000e+00  0.000000e+00   
org_id                      2.900000e+01  1.080000e+02  2.382500e+02   
invited_by_user_id          3.058000e+03  5.954000e+03  8.817000e+03   

                                     max  
object_id                   1.200000e+04  
last_session_creation_time  1.402067e+09  
opted_in_to_mailing_list    1.000000e+00  
enabled_for_marketing_drip  1.000000e+00  
org_id                      4.160000e+02  
invited_by_user_id          1.199900e+04

In [ ]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   object_id                   12000 non-null  int64  
 1   creation_time               12000 non-null  object 
 2   name                        12000 non-null  object 
 3   email                       12000 non-null  object 
 4   creation_source             12000 non-null  object 
 5   last_session_creation_time  8823 non-null   float64
 6   opted_in_to_mailing_list    12000 non-null  int64  
 7   enabled_for_marketing_drip  12000 non-null  int64  
 8   org_id                      12000 non-null  int64  
 9   invited_by_user_id          6417 non-null   float64
dtypes: float64(2), int64(4), object(4)
memory usage: 937.6+ KB


In [ ]:
users.isnull().sum()

object_id                        0
creation_time                    0
name                             0
email                            0
creation_source                  0
last_session_creation_time    3177
opted_in_to_mailing_list         0
enabled_for_marketing_drip       0
org_id                           0
invited_by_user_id            5583
dtype: int64

In [ ]:
users.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
11995    False
11996    False
11997    False
11998    False
11999    False
Length: 12000, dtype: bool

In [ ]:
user_engagement=pd.read_csv('/content/takehome_user_engagement.csv')
user_engagement.head()

time_stamp  user_id  visited
0  2014-04-22 03:53:30        1        1
1  2013-11-15 03:45:04        2        1
2  2013-11-29 03:45:04        2        1
3  2013-12-09 03:45:04        2        1
4  2013-12-25 03:45:04        2        1

In [ ]:
user_engagement.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207917 entries, 0 to 207916
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   time_stamp  207917 non-null  object
 1   user_id     207917 non-null  int64 
 2   visited     207917 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 4.8+ MB


In [ ]:
user_engagement.describe().T

count         mean          std  min     25%     50%     75%  \
user_id  207917.0  5913.314197  3394.941674  1.0  3087.0  5682.0  8944.0   
visited  207917.0     1.000000     0.000000  1.0     1.0     1.0     1.0   

             max  
user_id  12000.0  
visited      1.0

In [ ]:
user_engagement.user_id.nunique()

8823

In [ ]:
users.head()

object_id        creation_time               name  \
0          1  2014-04-22 03:53:30     Clausen August   
1          2  2013-11-15 03:45:04      Poole Matthew   
2          3  2013-03-19 23:14:52  Bottrill Mitchell   
3          4  2013-05-21 08:09:28    Clausen Nicklas   
4          5  2013-01-17 10:14:20          Raw Grace   

                        email creation_source  last_session_creation_time  \
0    AugustCClausen@yahoo.com    GUEST_INVITE                1.398139e+09   
1      MatthewPoole@gustr.com      ORG_INVITE                1.396238e+09   
2  MitchellBottrill@gustr.com      ORG_INVITE                1.363735e+09   
3   NicklasSClausen@yahoo.com    GUEST_INVITE                1.369210e+09   
4          GraceRaw@yahoo.com    GUEST_INVITE                1.358850e+09   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         1                           0      11   
1                         0                           0       1   
2                         0                           0      94   
3                         0                           0       1   
4                         0                           0     193   

   invited_by_user_id  
0             10803.0  
1               316.0  
2              1525.0  
3              5151.0  
4              5240.0

In [ ]:
user_engagement.head()

time_stamp  user_id  visited
0  2014-04-22 03:53:30        1        1
1  2013-11-15 03:45:04        2        1
2  2013-11-29 03:45:04        2        1
3  2013-12-09 03:45:04        2        1
4  2013-12-25 03:45:04        2        1

user_engagement dataframe seems to be sorted according to the user_id and time_stamp column and there is data for 8823 users out of the total 12000 users registered under the users dataframe.

In [ ]:
users['last_session_creation_time'] = pd.to_datetime(users['last_session_creation_time'] ,unit='s')

In [ ]:
users.last_session_creation_time.dtypes

dtype('<M8[ns]')

In [ ]:
users.head()

object_id        creation_time               name  \
0          1  2014-04-22 03:53:30     Clausen August   
1          2  2013-11-15 03:45:04      Poole Matthew   
2          3  2013-03-19 23:14:52  Bottrill Mitchell   
3          4  2013-05-21 08:09:28    Clausen Nicklas   
4          5  2013-01-17 10:14:20          Raw Grace   

                        email creation_source last_session_creation_time  \
0    AugustCClausen@yahoo.com    GUEST_INVITE        2014-04-22 03:53:30   
1      MatthewPoole@gustr.com      ORG_INVITE        2014-03-31 03:45:04   
2  MitchellBottrill@gustr.com      ORG_INVITE        2013-03-19 23:14:52   
3   NicklasSClausen@yahoo.com    GUEST_INVITE        2013-05-22 08:09:28   
4          GraceRaw@yahoo.com    GUEST_INVITE        2013-01-22 10:14:20   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         1                           0      11   
1                         0                           0       1   
2                         0                           0      94   
3                         0                           0       1   
4                         0                           0     193   

   invited_by_user_id  
0             10803.0  
1               316.0  
2              1525.0  
3              5151.0  
4              5240.0

In [ ]:
users['last_session_creation_time'].min(), users['last_session_creation_time'].max()

(Timestamp('2012-05-31 08:20:06'), Timestamp('2014-06-06 14:58:50'))

In [ ]:
users['creation_time'].min(), users['creation_time'].max()

('2012-05-31 00:43:27', '2014-05-30 23:59:19')

In [1]:
def get_date_int(df, column):
  
    year = df[column].dt.year
    month = df[column].dt.month
    day = df[column].dt.day
    return year, month, day

def get_week(x): return x.isocalendar()

def get_iso_date_int(df,column):
    '''
    With time coded as iso (year,week,day) this seperates those time periods.
    '''
    temp_df=pd.DataFrame(df[column].tolist(), index=df.index)
    year,week,day=temp_df[0],temp_df[1],temp_df[2]
    return year,week,day

In [ ]:
users = pd.read_csv('takehome_users.csv',encoding='ISO-8859-1')
#code creation_time,last_session_time as date/time
users.creation_time = pd.to_datetime(users['creation_time'])
users.last_session_creation_time = pd.to_datetime(users['last_session_creation_time'])
#change column heading
users['user_id'] = users['object_id']
#drop original column
users.drop('object_id', axis=1, inplace=True)
#drop private information
users.drop(['name', 'email'], axis=1, inplace=True)

users.head()

creation_time creation_source    last_session_creation_time  \
0 2014-04-22 03:53:30    GUEST_INVITE 1970-01-01 00:00:01.398138810   
1 2013-11-15 03:45:04      ORG_INVITE 1970-01-01 00:00:01.396237504   
2 2013-03-19 23:14:52      ORG_INVITE 1970-01-01 00:00:01.363734892   
3 2013-05-21 08:09:28    GUEST_INVITE 1970-01-01 00:00:01.369210168   
4 2013-01-17 10:14:20    GUEST_INVITE 1970-01-01 00:00:01.358849660   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         1                           0      11   
1                         0                           0       1   
2                         0                           0      94   
3                         0                           0       1   
4                         0                           0     193   

   invited_by_user_id  user_id  
0             10803.0        1  
1               316.0        2  
2              1525.0        3  
3              5151.0        4  
4              5240.0        5

In [ ]:
#make user engagement time_stamp into date/time
user_engagement['time_stamp'] = pd.to_datetime(user_engagement['time_stamp'])
#Make a weekly unit for time stamp
user_engagement['week_time_stamp']=user_engagement['time_stamp'].apply(get_week)

In [ ]:
print('First user engagement timestamp:',min(user_engagement.time_stamp))
print('Last user engagement timestamp:',max(user_engagement.time_stamp))

First user engagement timestamp: 2012-05-31 08:20:06
Last user engagement timestamp: 2014-06-06 14:58:50


In [ ]:
year, month, day=get_date_int(user_engagement, 'time_stamp')
user_engagement['year'],user_engagement['month'],user_engagement['day']=year,month,day
user_engagement['week']=user_engagement['time_stamp'].dt.week

iso_year,iso_week,iso_day=get_iso_date_int(user_engagement,'week_time_stamp')
user_engagement['year_week']=list(zip(iso_year,iso_week))

In [ ]:
user_engagement=user_engagement.sort_values(['time_stamp','user_id'],ascending=True)
user_engagement=user_engagement[['user_id','visited','day','year_week']]